Portal API

Zidane opened the pull request.

Zidane committed:
A Portal API.

This consolidates the old PortalAgent system into something that is actually implementable for Vanilla Minecraft but doesn't reduce flexiblity for plugin developers. Plugin devs are welcome to implement PortalType, register it, and use it to teleport entities however they please.

Additional features in this commit:

- Ability to send a viewer an environment, such as THE_NETHER. Allows a client to see a red skybox and fog.
- Ability to set the dimemsion type of a WorldProperties. This is similar to sending the environment to a viewer but is permanent and **may** change the logic of a World.
- Various other API cleanup mostly dealing with world changes

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
PortalType javadocs.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

gabizou commented on src/main/java/org/spongepowered/api/effect/Viewer.java:
```diff
@@ -46,6 +48,17 @@
+     * Sends the effect of being in a particular Vanilla world environment, such as the Nether,
+     * as an effect to the viewer.
+     *
+     * <p>For example, specifying {@link DimensionTypes#THE_NETHER}</p> will create a red skybox and
```
Why close the paragraph mid sentence?

Zidane replied:
Whoops missed that.

gabizou reviewed (approved):
A small nitpick

Cybermaxke commented on src/main/java/org/spongepowered/api/world/portal/PortalTypes.java:
```diff
@@ -22,19 +22,20 @@
 
-    // The default agent used to handle traveling to nether and the end
-    public static final Supplier<PortalAgentType> DEFAULT = Sponge.getRegistry().getCatalogRegistry().provideSupplier(PortalAgentType.class, "DEFAULT");
+    public static final Supplier<PortalType> NETHER = Sponge.getRegistry().getCatalogRegistry().provideSupplier(PortalType.class, "NETHER");
```
`THE_END` and `THE_NETHER` to be consistent with dimension and generator types naming?

dualspiral replied:
I don't think that makes sense here. We generate "the nether" and "the end" (there is one nether dimension type and one nether generator, for example) - but we use "nether" and "end" portals, not "the nether" or "the end" portals - of which there may be multiple. I think it's correct as is.

dualspiral commented on src/main/java/org/spongepowered/api/event/entity/MoveEntityEvent.java:
```diff
@@ -70,102 +72,14 @@
     /**
-     * Fired when an {@link Entity}'s position changes for reasons other than
-     * normal movement.
+     * Fired when an {@link Entity entity's} position goes across {@link ServerWorld worlds}.
```
I don't think it should really be named "Teleport" if this is the condition, because plugin devs associate a teleport as a non-natural, such as `/tp`, movement, and they often don't involve a world change.

A `ChangeWorldEvent` is a specialised `MoveEntityEvent` really. I would argue that if you are changing world, you should _always_ be teleporting to a specific position meaning a standalone event makes no sense. So I think that should be a sub-event, and add a `MovementType` to the event itself so plugins can inspect what type of movement caused this.

Zidane replied:
There is the problem though.

Portal repositioning logic is VERY expensive. I fire the ChangeEntityWorld right away and allow you to change the world.

Not only that but I can't allow the Teleport to set the world as it breaks' the Mojang Teleporter so I compromised by letting you do that here (in-case you only care about the World you are going to) and then if the dev cares about the position you must cancel and do it yourself.

dualspiral commented on src/main/java/org/spongepowered/api/event/cause/entity/teleport/MovementTypes.java:
```diff
@@ -28,24 +28,24 @@
-    public static final Supplier<TeleportType> PORTAL = Sponge.getRegistry().getCatalogRegistry().provideSupplier(TeleportType.class, "PORTAL");
+    public static final Supplier<MovementType> PLUGIN = Sponge.getRegistry().getCatalogRegistry().provideSupplier(MovementType.class, "PLUGIN");
 
-    public static final Supplier<TeleportType> UNKNOWN = Sponge.getRegistry().getCatalogRegistry().provideSupplier(TeleportType.class, "UNKNOWN");
```
Is there a situation where we won't be able to determine what the movement type was? We might want to keep `UNKNOWN` if that's the case.

Zidane replied:
If it is unknown it should be treated as plugin. We control exactly when this is used so we can set the movement. Unknown, by the merits, is unhelpful.

dualspiral commented on src/main/java/org/spongepowered/api/event/entity/MoveEntityEvent.java:
```diff
@@ -70,102 +72,14 @@
      */
     @GenerateFactoryMethod
-    interface Position extends MoveEntityEvent {}
+    interface Position extends MoveEntityEvent {
```
This was added to be able to separate position and rotation movement - but the rotation event is actually separate. So why do we need this? It would make more sense if this event contained entity rotation... but it does not.

I think we should specify that the `MoveEntityEvent` refers to position only, and that the rotation event is rotation only and just separate the movement and rotation logic completely.

dualspiral commented on src/main/java/org/spongepowered/api/event/entity/MoveEntityEvent.java:
```diff
@@ -27,36 +27,38 @@
  * Called when an {@link Entity} performs movement.
  */
-@GenerateFactoryMethod
 public interface MoveEntityEvent extends Event, Cancellable {
```
I think this should contain `getServerWorld()` or `getServerLocation()` as many plugins will be interested in the entity's location, which includes the world they are in, and will want to act upon that. Don't need a setter, but a getter will support plugin devs better.

dualspiral reviewed (commented):
Mostly have thoughts about the events - I think they need a bit of rejigging.

Zidane committed:
Fixes from feedback. Javadoc corrections.

- Remove Dimension from API, was an implementation detail
- Fix command parameters specifying a "DIMENSION" parameter which would have never worked correctly. In MCP, Dimensions are 1 per world and are not singleton. Even worse, we were blindly casting, in MCP, to Mojang's DimensionType which is a different concept altogether...

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Slight corrections to fit our style.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Add MovementTypes END_GATEWAY, ENTITY_TELEPORT

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Cybermaxke commented on src/main/java/org/spongepowered/api/event/cause/entity/teleport/MovementTypes.java:
```diff
@@ -28,24 +28,28 @@
+    public static final Supplier<MovementType> ENDER_PEARL = Sponge.getRegistry().getCatalogRegistry().provideSupplier(MovementType.class, "ENDER_PEARL");
 
-    public static final Supplier<TeleportType> PORTAL = Sponge.getRegistry().getCatalogRegistry().provideSupplier(TeleportType.class, "PORTAL");
+    public static final Supplier<MovementType> ENTITY_TELEPORT = Sponge.getRegistry().getCatalogRegistry().provideSupplier(MovementType.class, "ENTITY_TELEPORT");
```
Just `TELEPORT`?

Zidane committed:
Re-structure around where we store Cause and where the cause objects are stored in the API.

This is cleaner, no need for cause.Cause...should be on the base event as all events have Causes.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

dualspiral commented:
7bc9064cf5c2e23cac42bcfcb8d98c1f0841b6e7
